server/resp/src/main/java/org/infinispan/server/resp/commands/list/blocking/BLPOP.java

```diff
@@ -0,0 +1,205 @@
+
+/**
+ * @link https://redis.io/commands/blpop/
+ * @since 15.0
+ */
+public class BLPOP extends RespCommand implements Resp3Command {
+   private static final Log log = LogFactory.getLog(MethodHandles.lookup().lookupClass(), Log.class);
+
+   public BLPOP() {
+      super(-3, 1, -2, 1);
+   }
+
+   @Override
+   public CompletionStage<RespRequestHandler> perform(Resp3Handler handler,
+         ChannelHandlerContext ctx,
+         List<byte[]> arguments) {
+      EmbeddedMultimapListCache<byte[], byte[]> listMultimap = handler.getListMultimap();
+      var lastKeyIdx = arguments.size() - 1;
+      var filterKeys = arguments.subList(0, lastKeyIdx);
+      // Using last arg as timeout if it can be a double
+      var argTimeout = ArgumentUtils.toDouble(arguments.get(lastKeyIdx));
+      if (argTimeout < 0) {
+         RespErrorUtil.mustBePositive(handler.allocator());
+         return handler.myStage();
+      }
+      long timeout = (long) (argTimeout * Duration.ofSeconds(1).toMillis());
+
+      // If all the keys are empty or null, create a listener for each key
+      // otherwise return the left value of the first non empty list
+      var pollStage = pollAllKeys(listMultimap, arguments);
+      // If no value returned, we need subscribers
+      return handler.stageToReturn(pollStage.thenCompose(v -> {
+         // addSubscriber call can rise exception that needs to be reported
+         // as error
+            var retStage = (v != null && !v.isEmpty())
+                  ? CompletableFuture.completedFuture(v)
+                  : addSubscriber(listMultimap, filterKeys, timeout, handler, ctx);
+                  return retStage;
+      }), ctx, Consumers.COLLECTION_BULK_BICONSUMER);
+   }
+
+   CompletionStage<Collection<byte[]>> pollKeyValue(EmbeddedMultimapListCache<byte[], byte[]> mmList, byte[] key) {
+      return mmList.pollFirst(key, 1)
+            .thenApply((v) -> (v == null || v.isEmpty())
+                  ? null
+                  : Arrays.asList(key, v.iterator().next()));
+   }
+
+   CompletionStage<Collection<byte[]>> addSubscriber(EmbeddedMultimapListCache<byte[], byte[]> listMultimap,
+         List<byte[]> filterKeys, long timeout, Resp3Handler handler,
+         ChannelHandlerContext ctx) {
+      if (log.isTraceEnabled()) {
+         log.tracef("Subscriber for keys: " +
+               filterKeys.toString());
+      }
+      AdvancedCache<Object, Object> cache = handler.cache().withMediaType(MediaType.APPLICATION_OCTET_STREAM, null);
+      DataConversion vc = cache.getValueDataConversion();
+      PubSubListener pubSubListener = new PubSubListener(ctx.channel());
+      EventListenerKeysFilter filter = new EventListenerKeysFilter(filterKeys.toArray(byte[][]::new));
+      CacheEventConverter<Object, Object, Object> converter = (key, oldValue, oldMetadata, newValue, newMetadata,
+            eventType) -> vc.fromStorage(newValue);
+      cache.addListener(pubSubListener, filter, converter);
```

Comment:
@wburns Could this still block the event loop?

Reply:
With the listener CF as `CompletionStage<Void> listener = cache.addListenerAsync(pubSubListener, filter, converter);`, what I thought originally was:

```java
      CompletionStage<Void> listener = cache.addListenerAsync(pubSubListener, filter, converter);
      // ... Code handling timeout and pooling again...
      return CompletionStages.handleAndCompose(listener, (ignore, t) -> {
         if (t != null) {
            pubSubListener.getFuture().completeExceptionally(t);
         }

         return pubSubListener.future;
      }).thenApply(ret -> {
         if (scheduledTimer != null) scheduledTimer.cancel(true);
         return ret;
      });
```

Reply:
nice! applied with some changes

Reply:
> @wburns Could this still block the event loop?

Just to close this up, yes it definitely would. I would have expected blockhound to detect this when you have > 1 nodes btw.